### PR TITLE
Add tests for toggle stateful assist

### PR DESCRIPTION
**Pull request by joebingham-wk**
Just added the tests specifically for the ToggleStatefulAssist!

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+  @override
+  AssistKind get assistKindUnderTest => ToggleComponentStatefulness.makeStateful;
+
+  Future<void> test_noAssistOnFactory() async {
+    final generatedSource = simpleUiComponentSource();
+
+    var source = newSource(BoilerplateAssistTestStrings.fileName, generatedSource);
+    var selection = createSelection(source,
+        generatedSource.replaceFirst(r'UiFactory<FooProps> Foo = _$Foo;', r'UiFactory<FooProps> #Foo# = _$Foo;'));
```
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart:30]: Why the replacement and not just the following?
```suggestion
    var selection = createSelection(source, r'UiFactory<FooProps> #Foo# = _$Foo;');
```
The whole source isn't needed to make a selection; the context before/after can be as small/large as you want

File: tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+class AddStatefulnessAssist extends AssistTestBase with BoilerplateAssistTestStrings {
+  static int expectedPriority = ToggleComponentStatefulness.statefulAssistPriority;
+
+  @override
+  AssistKind get assistKindUnderTest => ToggleComponentStatefulness.makeStateful;
+
+  Future<void> test_noAssistOnFactory() async {
+    final generatedSource = simpleUiComponentSource();
+
+    var source = newSource(BoilerplateAssistTestStrings.fileName, generatedSource);
```
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart:28]: #nit is having this long constant necessary? I would say just put in a dummy name, or make `fileName` an instance member so you don't have to prefix it
```suggestion
    var source = newSource(fileName, generatedSource);
```

Comment (greglittlefield-wf) [reply tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart:28]: We could also just make newSource's fileName arg be an optional/named arg, since a lot of the time you don't beed it
```suggestion
    var source = newSource(generatedSource);
```

File: tools/analyzer_plugin/test/integration/assists/boilerplate_assist_utils.dart [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+  String componentName({bool withSelection = false}) => withSelection ? '#${prefix}Component#' : '${prefix}Component';
+
+  String simpleComponentSuperclass({bool withState = false}) =>
+      withState ? 'UiStatefulComponent2<${prefix}Props, ${prefix}State>' : 'UiComponent2<${prefix}Props>';
+
+  String fluxComponentSuperclass({bool withState = false}) =>
+      withState ? 'FluxUiStatefulComponent2<${prefix}Props, ${prefix}State>' : 'FluxUiComponent2<${prefix}Props>';
+
+  String simpleUiComponentSource({bool isStateful = false, bool shouldIncludeSelection = false}) {
```
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/test/integration/assists/boilerplate_assist_utils.dart:25]: >shouldIncludeSelection

See my note below about `#`

File: tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    await expectNoAssist(selection);
+  }
+
+  Future<void> test_addsStatefulness() async {
+    var retrievedSource = newSource(BoilerplateAssistTestStrings.fileName, simpleUiComponentSource());
+    var selection = createSelection(retrievedSource, simpleUiComponentSource(shouldIncludeSelection: true));
+    final change = await expectAndGetSingleAssist(selection);
+    retrievedSource = applySourceChange(change, retrievedSource);
+    expect(retrievedSource.contents.data, simpleUiComponentSource(isStateful: true));
+  }
```
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart:72]: Since there's logic around the presence of `defaultProps`, can we add a test case for when it's not present?

File: tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+  Future<void> test_removesStatefulness() async {
+    var retrievedSource = newSource(BoilerplateAssistTestStrings.fileName, simpleUiComponentSource(isStateful: true));
+    var selection =
+        createSelection(retrievedSource, simpleUiComponentSource(shouldIncludeSelection: true, isStateful: true));
+    final change = await expectAndGetSingleAssist(selection);
+    retrievedSource = applySourceChange(change, retrievedSource);
+    expect(retrievedSource.contents.data, simpleUiComponentSource());
+  }
```
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart:141]: Since there's logic to conditionally remove `initialState`, can we add a test case for when it's not present?

File: tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+class ${withSelection ? '#TestComponent#' : 'TestComponent'} extends UiComponent<TestProps> {
+  @override
+  render() {}
+}
+''';
+
+    var source = newSource(BoilerplateAssistTestStrings.fileName, oldBoilerplate());
+    var selection = createSelection(source, oldBoilerplate(withSelection: true));
+    await expectNoAssist(selection);
+  }
```
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart:64]: Can we add a couple test cases for edge cases?

- non-supported component base class
- running this assist in a file with no component

File: tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart (hunk 2 above)
Comment (joebingham-wk) [reply tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart:28]: Good call, I can shorten that! We could make it optional, but I think that's a larger question than this PR since the method is in use elsewhere (and those spots would have to be updated too)

Comment (greglittlefield-wf) [reply tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart:28]: Oh yeah for sure, we could definitely address that separately; just thought it'd be worth mentioning that these helper methods are still fairly new and we should feel free to change em to better suit our use cases!

Comment (joebingham-wk) [reply tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart:28]: True true, that's a good point!

File: tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart (hunk 6 above)
Comment (joebingham-wk) [reply tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart:64]: For sure! How would the non-supported base class differ from https://github.com/Workiva/over_react/pull/580/files#diff-a479f45e42c16ab64749a5016a13af57R43? Is there another base class we should test against?

But added a test with a file with no component!

Comment (greglittlefield-wf) [reply tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart:64]: My thought process was to test `_getNewBase ` and the related code path: https://github.com/Workiva/over_react/blob/master/tools/analyzer_plugin/lib/src/assist/toggle_stateful.dart#L49-L52

since for old boilerplate, the assist will short-circuit by returning false in `initializeAssistApi`, since `ComponentDeclarationAssistApi` doesn't support legacy boilerplate

Comment (joebingham-wk) [reply tools/analyzer_plugin/test/integration/assists/toggle_statefulness_test.dart:64]: Ohh, duhh, haha. Good call, adding that in the other branch!
